messaging/src/main/java/nl/nn/adapterframework/extensions/kafka/KafkaListener.java

```diff
@@ -0,0 +1,197 @@
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Predicate;
+import java.util.regex.Pattern;
+
+import javax.annotation.Nonnull;
+
+import org.apache.commons.lang3.StringUtils;
+import org.apache.kafka.clients.consumer.Consumer;
+import org.apache.kafka.clients.consumer.ConsumerConfig;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.consumer.KafkaConsumer;
+import org.apache.kafka.clients.consumer.OffsetAndMetadata;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.serialization.ByteArrayDeserializer;
+import org.apache.kafka.common.serialization.StringDeserializer;
+
+import lombok.AccessLevel;
+import lombok.Getter;
+import lombok.Setter;
+import nl.nn.adapterframework.configuration.ConfigurationException;
+import nl.nn.adapterframework.core.IListener;
+import nl.nn.adapterframework.core.IPullingListener;
+import nl.nn.adapterframework.core.ListenerException;
+import nl.nn.adapterframework.core.PipeLineResult;
+import nl.nn.adapterframework.core.PipeLineSession;
+import nl.nn.adapterframework.receivers.RawMessageWrapper;
+import nl.nn.adapterframework.stream.Message;
+import nl.nn.adapterframework.stream.MessageContext;
+import nl.nn.adapterframework.util.StringUtil;
+
+/**
+ * Experimental {@link IListener} for listening to a topic in
+ * a Kafka instance.
+ */
+public class KafkaListener extends KafkaFacade implements IPullingListener<ConsumerRecord<String, byte[]>> {
+
+	private static final Predicate<String> TOPIC_NAME_PATTERN = Pattern.compile("^[a-zA-Z0-9._\\-*]*$").asPredicate();
+
+	/** The group id of the consumer */
+	private @Setter String groupId;
+
+	/** Whether to start reading from the beginning of the topic. */
+	private @Setter boolean fromBeginning = false;
+
+	/** How often to check for new topics when using Patterns. (in MS) */
+	private @Setter int patternRecheckInterval = 5000;
+
+	/**
+	 * The topics to listen to as comma-separated list. Regular expressions are supported,
+	 * for instance: {@code example.*}.
+	 */
+	private @Setter String topics;
+
+	//setter is for testing purposes only.
+	private @Setter(AccessLevel.PACKAGE) Consumer<String, byte[]> consumer;
+	private Iterator<? extends ConsumerRecord<String, byte[]>> waiting;
+	private final Duration pollDuration = Duration.ofMillis(1);
+	private final Map<TopicPartition, OffsetAndMetadata> offsetAndMetadataMap = new HashMap<>();
+	private @Getter(AccessLevel.PACKAGE) Pattern topicPattern;
+	private final Lock lock = new ReentrantLock();
+
+	@Override
+	public void configure() throws ConfigurationException {
+		super.configure();
+		if (StringUtils.isEmpty(groupId)) throw new ConfigurationException("groupId must be specified");
+		if (StringUtils.isEmpty(topics)) throw new ConfigurationException("topics must be specified");
+		if(patternRecheckInterval < 10) throw new ConfigurationException("patternRecheckInterval should be at least 10");
```

Comment:
Can you apply code formatting on the whole class?

Reply:
Doesn't change these lines on my PC... 😬

Reply:
Oh strange, I would suspect a whitespace in the if-construct. At least, that's what my IJ does. Perhaps apply it twice or so (also not the changed lines / whole file). I'll close it too.